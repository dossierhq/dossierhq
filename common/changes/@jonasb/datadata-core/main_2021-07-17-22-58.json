{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Add OkResult.map() and ErrorType.isErrorType()",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}