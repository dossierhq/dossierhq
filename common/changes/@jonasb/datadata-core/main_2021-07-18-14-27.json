{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Configure importHelpers for ts, and add tslib dependency",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}