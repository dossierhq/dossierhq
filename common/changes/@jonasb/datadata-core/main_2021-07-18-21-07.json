{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Export transpiled JavaScript",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}