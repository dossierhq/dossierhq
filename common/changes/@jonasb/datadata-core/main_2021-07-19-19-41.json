{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Fix export configuration",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}