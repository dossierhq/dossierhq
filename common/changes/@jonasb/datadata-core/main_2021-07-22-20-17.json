{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Allow string in schema field specification",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}