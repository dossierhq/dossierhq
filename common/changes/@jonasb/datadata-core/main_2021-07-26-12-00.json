{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Enable easier JSON serialization of adminClient",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}