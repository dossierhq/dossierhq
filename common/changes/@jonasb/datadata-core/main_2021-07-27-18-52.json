{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Support boolean type",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}