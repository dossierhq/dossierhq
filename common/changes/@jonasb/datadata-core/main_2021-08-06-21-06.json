{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "upsertEntity() support in AdminClient",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}