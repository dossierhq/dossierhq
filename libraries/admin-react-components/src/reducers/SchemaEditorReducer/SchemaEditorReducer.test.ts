import { AdminSchema, FieldType } from '@jonasb/datadata-core';
import type { SchemaEditorState, SchemaEditorStateAction } from './SchemaEditorReducer';
import {
  initializeSchemaEditorState,
  reduceSchemaEditorState,
  SchemaEditorActions,
} from './SchemaEditorReducer';

function reduceSchemaEditorStateActions(
  state: SchemaEditorState,
  ...actions: SchemaEditorStateAction[]
) {
  let newState = state;
  for (const action of actions) {
    newState = reduceSchemaEditorState(newState, action);
  }
  return newState;
}

describe('initializeSchemaEditorState', () => {
  test('no args', () => {
    expect(initializeSchemaEditorState()).toMatchInlineSnapshot(`
      Object {
        "entityTypes": Array [],
        "schema": null,
        "valueTypes": Array [],
      }
    `);
  });
});

describe('AddEntityTypeAction', () => {
  test('add type', () => {
    const state = reduceSchemaEditorState(
      initializeSchemaEditorState(),
      new SchemaEditorActions.AddEntityType('Foo')
    );
    expect(state).toMatchInlineSnapshot(`
      Object {
        "entityTypes": Array [
          Object {
            "fields": Array [],
            "name": "Foo",
            "type": "entity",
          },
        ],
        "schema": null,
        "valueTypes": Array [],
      }
    `);
  });

  test('add two types (orders)', () => {
    const state = reduceSchemaEditorStateActions(
      initializeSchemaEditorState(),
      new SchemaEditorActions.AddEntityType('ZooKeeper'),
      new SchemaEditorActions.AddEntityType('Anaconda')
    );
    expect(state).toMatchInlineSnapshot(`
      Object {
        "entityTypes": Array [
          Object {
            "fields": Array [],
            "name": "Anaconda",
            "type": "entity",
          },
          Object {
            "fields": Array [],
            "name": "ZooKeeper",
            "type": "entity",
          },
        ],
        "schema": null,
        "valueTypes": Array [],
      }
    `);
  });
});

describe('UpdateSchemaSpecificationAction', () => {
  test('add empty schema', () => {
    const emptySchemaState = reduceSchemaEditorState(
      initializeSchemaEditorState(),
      new SchemaEditorActions.UpdateSchemaSpecification(
        new AdminSchema({ entityTypes: [], valueTypes: [] })
      )
    );
    expect(emptySchemaState).toMatchInlineSnapshot(`
      Object {
        "entityTypes": Array [],
        "schema": AdminSchema {
          "spec": Object {
            "entityTypes": Array [],
            "valueTypes": Array [],
          },
        },
        "valueTypes": Array [],
      }
    `);
  });

  test('one entity type', () => {
    const schemaState = reduceSchemaEditorState(
      initializeSchemaEditorState(),
      new SchemaEditorActions.UpdateSchemaSpecification(
        new AdminSchema({
          entityTypes: [
            {
              name: 'TitleOnly',
              adminOnly: false,
              fields: [{ name: 'title', type: FieldType.String }],
            },
          ],
          valueTypes: [],
        })
      )
    );
    expect(schemaState).toMatchInlineSnapshot(`
      Object {
        "entityTypes": Array [
          Object {
            "fields": Array [
              Object {
                "list": false,
                "name": "title",
                "type": "String",
              },
            ],
            "name": "TitleOnly",
            "type": "entity",
          },
        ],
        "schema": AdminSchema {
          "spec": Object {
            "entityTypes": Array [
              Object {
                "adminOnly": false,
                "fields": Array [
                  Object {
                    "name": "title",
                    "type": "String",
                  },
                ],
                "name": "TitleOnly",
              },
            ],
            "valueTypes": Array [],
          },
        },
        "valueTypes": Array [],
      }
    `);
  });

  test('one value type', () => {
    const schemaState = reduceSchemaEditorState(
      initializeSchemaEditorState(),
      new SchemaEditorActions.UpdateSchemaSpecification(
        new AdminSchema({
          entityTypes: [],
          valueTypes: [
            {
              name: 'TitleOnly',
              adminOnly: false,
              fields: [{ name: 'title', type: FieldType.String }],
            },
          ],
        })
      )
    );
    expect(schemaState).toMatchInlineSnapshot(`
      Object {
        "entityTypes": Array [],
        "schema": AdminSchema {
          "spec": Object {
            "entityTypes": Array [],
            "valueTypes": Array [
              Object {
                "adminOnly": false,
                "fields": Array [
                  Object {
                    "name": "title",
                    "type": "String",
                  },
                ],
                "name": "TitleOnly",
              },
            ],
          },
        },
        "valueTypes": Array [
          Object {
            "fields": Array [
              Object {
                "list": false,
                "name": "title",
                "type": "String",
              },
            ],
            "name": "TitleOnly",
            "type": "value",
          },
        ],
      }
    `);
  });
});
