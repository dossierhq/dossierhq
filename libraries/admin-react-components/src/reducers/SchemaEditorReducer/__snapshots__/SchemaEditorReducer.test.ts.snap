// Vitest Snapshot v1

exports[`ChangeFieldAdminOnlyAction > make new field admin only in existing entity type 1`] = `
{
  "activeSelector": null,
  "activeSelectorEditorScrollSignal": 0,
  "activeSelectorMenuScrollSignal": 0,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": true,
          "existingFieldSpec": null,
          "isName": false,
          "list": false,
          "multiline": false,
          "name": "bar",
          "required": false,
          "status": "new",
          "type": "String",
        },
      ],
      "kind": "entity",
      "name": "Foo",
      "status": "changed",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [
        {
          "adminOnly": false,
          "authKeyPattern": null,
          "fields": [],
          "name": "Foo",
        },
      ],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`ChangeFieldAdminOnlyAction > make new field admin only in existing entity type 2`] = `
{
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": true,
          "multiline": false,
          "name": "bar",
          "required": false,
          "type": "String",
        },
      ],
      "name": "Foo",
    },
  ],
}
`;

exports[`ChangeFieldAllowedLinkEntityTypesAction > change link entity types of a new entity field 1`] = `
{
  "activeSelector": null,
  "activeSelectorEditorScrollSignal": 0,
  "activeSelectorMenuScrollSignal": 0,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "existingFieldSpec": null,
          "isName": false,
          "linkEntityTypes": [
            "Foo",
          ],
          "list": false,
          "name": "foo",
          "required": false,
          "richTextNodes": [],
          "status": "new",
          "type": "RichText",
        },
      ],
      "kind": "entity",
      "name": "Foo",
      "status": "changed",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [
        {
          "adminOnly": false,
          "authKeyPattern": null,
          "fields": [],
          "name": "Foo",
        },
      ],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`ChangeFieldAllowedLinkEntityTypesAction > change link entity types of a new entity field 2`] = `
{
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "entityTypes": [],
          "linkEntityTypes": [
            "Foo",
          ],
          "name": "foo",
          "required": false,
          "richTextNodes": [],
          "type": "RichText",
          "valueTypes": [],
        },
      ],
      "name": "Foo",
    },
  ],
}
`;

exports[`ChangeFieldIsNameAction > make new string field is-name in new type 1`] = `
{
  "activeSelector": {
    "kind": "entity",
    "typeName": "TitleOnly",
  },
  "activeSelectorEditorScrollSignal": 1,
  "activeSelectorMenuScrollSignal": 1,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "existingFieldSpec": null,
          "isName": true,
          "list": false,
          "multiline": false,
          "name": "title",
          "required": false,
          "status": "new",
          "type": "String",
        },
      ],
      "kind": "entity",
      "name": "TitleOnly",
      "status": "new",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`ChangeFieldIsNameAction > make new string field is-name in new type 2`] = `
{
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "isName": true,
          "multiline": false,
          "name": "title",
          "required": false,
          "type": "String",
        },
      ],
      "name": "TitleOnly",
    },
  ],
}
`;

exports[`ChangeFieldIsNameAction > make other field is-name, and switch back 1`] = `
{
  "activeSelector": null,
  "activeSelectorEditorScrollSignal": 0,
  "activeSelectorMenuScrollSignal": 0,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "existingFieldSpec": {
            "isName": true,
            "name": "title",
            "type": "String",
          },
          "isName": false,
          "list": false,
          "multiline": false,
          "name": "title",
          "required": false,
          "status": "changed",
          "type": "String",
        },
        {
          "adminOnly": false,
          "existingFieldSpec": null,
          "isName": true,
          "list": false,
          "multiline": false,
          "name": "other",
          "required": false,
          "status": "new",
          "type": "String",
        },
      ],
      "kind": "entity",
      "name": "TitleOnly",
      "status": "changed",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [
        {
          "adminOnly": false,
          "authKeyPattern": null,
          "fields": [
            {
              "isName": true,
              "name": "title",
              "type": "String",
            },
          ],
          "name": "TitleOnly",
        },
      ],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`ChangeFieldIsNameAction > make other field is-name, and switch back 2`] = `
{
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "multiline": false,
          "name": "title",
          "required": false,
          "type": "String",
        },
        {
          "adminOnly": false,
          "isName": true,
          "multiline": false,
          "name": "other",
          "required": false,
          "type": "String",
        },
      ],
      "name": "TitleOnly",
    },
  ],
}
`;

exports[`ChangeFieldIsNameAction > make other field is-name, and switch back 3`] = `
{
  "activeSelector": null,
  "activeSelectorEditorScrollSignal": 0,
  "activeSelectorMenuScrollSignal": 0,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "existingFieldSpec": {
            "isName": true,
            "name": "title",
            "type": "String",
          },
          "isName": true,
          "list": false,
          "multiline": false,
          "name": "title",
          "required": false,
          "status": "",
          "type": "String",
        },
        {
          "adminOnly": false,
          "existingFieldSpec": null,
          "isName": false,
          "list": false,
          "multiline": false,
          "name": "other",
          "required": false,
          "status": "new",
          "type": "String",
        },
      ],
      "kind": "entity",
      "name": "TitleOnly",
      "status": "changed",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [
        {
          "adminOnly": false,
          "authKeyPattern": null,
          "fields": [
            {
              "isName": true,
              "name": "title",
              "type": "String",
            },
          ],
          "name": "TitleOnly",
        },
      ],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`ChangeFieldMultilineAction > make new multiline field in existing entity type 1`] = `
{
  "activeSelector": null,
  "activeSelectorEditorScrollSignal": 0,
  "activeSelectorMenuScrollSignal": 0,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "existingFieldSpec": null,
          "isName": false,
          "list": false,
          "multiline": true,
          "name": "bar",
          "required": false,
          "status": "new",
          "type": "String",
        },
      ],
      "kind": "entity",
      "name": "Foo",
      "status": "changed",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [
        {
          "adminOnly": false,
          "authKeyPattern": null,
          "fields": [],
          "name": "Foo",
        },
      ],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`ChangeFieldMultilineAction > make new multiline field in existing entity type 2`] = `
{
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "multiline": true,
          "name": "bar",
          "required": false,
          "type": "String",
        },
      ],
      "name": "Foo",
    },
  ],
}
`;

exports[`RenameTypeAction > add and rename entity type with fields referring to itself 1`] = `
{
  "activeSelector": {
    "kind": "entity",
    "typeName": "Bar",
  },
  "activeSelectorEditorScrollSignal": 2,
  "activeSelectorMenuScrollSignal": 2,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "entityTypes": [
            "Bar",
          ],
          "existingFieldSpec": null,
          "isName": false,
          "linkEntityTypes": undefined,
          "list": false,
          "name": "self",
          "required": false,
          "status": "new",
          "type": "EntityType",
        },
      ],
      "kind": "entity",
      "name": "Bar",
      "status": "new",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`RenameTypeAction > add and rename entity type with fields referring to itself 2`] = `
{
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "entityTypes": [
            "Bar",
          ],
          "name": "self",
          "required": false,
          "type": "EntityType",
        },
      ],
      "name": "Bar",
    },
  ],
}
`;

exports[`RenameTypeAction > add and rename entity type with link fields referring to itself 1`] = `
{
  "activeSelector": {
    "kind": "entity",
    "typeName": "Bar",
  },
  "activeSelectorEditorScrollSignal": 2,
  "activeSelectorMenuScrollSignal": 2,
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "entityTypes": undefined,
          "existingFieldSpec": null,
          "isName": false,
          "linkEntityTypes": [
            "Bar",
          ],
          "list": false,
          "name": "self",
          "required": false,
          "richTextNodes": [],
          "status": "new",
          "type": "RichText",
        },
      ],
      "kind": "entity",
      "name": "Bar",
      "status": "new",
    },
  ],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [],
}
`;

exports[`RenameTypeAction > add and rename entity type with link fields referring to itself 2`] = `
{
  "entityTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "entityTypes": [],
          "linkEntityTypes": [
            "Bar",
          ],
          "name": "self",
          "required": false,
          "richTextNodes": [],
          "type": "RichText",
          "valueTypes": [],
        },
      ],
      "name": "Bar",
    },
  ],
}
`;

exports[`RenameTypeAction > add and rename value type with fields referring to itself 1`] = `
{
  "activeSelector": {
    "kind": "value",
    "typeName": "Bar",
  },
  "activeSelectorEditorScrollSignal": 2,
  "activeSelectorMenuScrollSignal": 2,
  "entityTypes": [],
  "schema": AdminSchema {
    "spec": {
      "entityTypes": [],
      "patterns": [],
      "valueTypes": [],
    },
  },
  "schemaWillBeUpdatedDueToSave": false,
  "status": "changed",
  "valueTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "existingFieldSpec": null,
          "isName": false,
          "list": false,
          "name": "self",
          "required": false,
          "status": "new",
          "type": "ValueType",
          "valueTypes": [
            "Bar",
          ],
        },
      ],
      "kind": "value",
      "name": "Bar",
      "status": "new",
    },
  ],
}
`;

exports[`RenameTypeAction > add and rename value type with fields referring to itself 2`] = `
{
  "valueTypes": [
    {
      "adminOnly": false,
      "fields": [
        {
          "adminOnly": false,
          "name": "self",
          "required": false,
          "type": "ValueType",
          "valueTypes": [
            "Bar",
          ],
        },
      ],
      "name": "Bar",
    },
  ],
}
`;
